common/workunit/workunit.cpp

```diff
@@ -1121,8 +1031,67 @@ class CDaliWorkUnit : public CPersistedWorkUnit
             throw;
         }
     }
+    virtual void setGraphState(const char *graphName, WUGraphState state) const
+    {
+        Owned<IRemoteConnection> conn = getWritableProgressConnection(graphName);
+        conn->queryRoot()->setPropInt("@_state", state);
+    }
+    virtual void setNodeState(const char *graphName, WUGraphIDType nodeId, WUGraphState state) const
+    {
+        CriticalBlock block(crit);
+        progressConnection.clear();  // Make sure nothing is locking for read or we won't be able to lock for write
+        VStringBuffer path("/GraphProgress/%s", queryWuid());
+        Owned<IRemoteConnection> conn = querySDS().connect(path, myProcessSession(), RTM_LOCK_WRITE|RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT);
```

Comment:
That's because this information is stored all over the place. I could lock the whole progress info each time I suppose. There's little benefit to caching the writeable connection, though perhaps keeping it the same as the read one and using changemode would be a slight efficiency gain (note - it's already much better than existing)